File: python/examples/synthetic_generation.py
#!/usr/bin/env python3
"""
Synthetic Data Generation with Normalizing Flows

This example demonstrates how to:
1. Train a flow on historical return data
2. Generate realistic synthetic scenarios
3. Use synthetic data for stress testing
4. Monte Carlo simulation for portfolio risk

Usage:
    python synthetic_generation.py
"""

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import numpy as np
import torch
import matplotlib.pyplot as plt
from scipy import stats

from flows import RealNVP
from data_fetcher import create_synthetic_returns, BybitDataFetcher
from training import FlowTrainer
from risk_metrics import stress_test, compute_var, compute_cvar


def main():
    print("=" * 60)
    print("Synthetic Data Generation with Normalizing Flows")
    print("=" * 60)

    # -----------------------------------------------------------------
    # Step 1: Prepare training data
    # -----------------------------------------------------------------
    print("\n[Step 1] Preparing training data...")

    try:
        fetcher = BybitDataFetcher()
        returns = fetcher.prepare_training_data(
            symbol="BTC/USDT",
            timeframe="1d",
            limit=500,
            standardize=False
        )
        data_source = "BTC/USDT"
    except Exception:
        print("  Using synthetic fat-tailed data...")
        returns = create_synthetic_returns(500, "fat_tails")
        data_source = "Synthetic"

    returns_flat = returns.flatten()
    print(f"  Data source: {data_source}")
    print(f"  Samples: {len(returns_flat)}")

    # -----------------------------------------------------------------
    # Step 2: Train normalizing flow
    # -----------------------------------------------------------------
    print("\n[Step 2] Training normalizing flow...")

    model = RealNVP(dim=1, n_layers=8, hidden_dim=128)
    trainer = FlowTrainer(model)

    trainer.train(
        train_data=returns,
        epochs=150,
        batch_size=64,
        verbose=True
    )

    # -----------------------------------------------------------------
    # Step 3: Generate synthetic scenarios
    # -----------------------------------------------------------------
    print("\n[Step 3] Generating synthetic return scenarios...")

    model.eval()
    with torch.no_grad():
        # Generate different amounts of synthetic data
        synthetic_1k = model.sample(1000).cpu().numpy().flatten()
        synthetic_10k = model.sample(10000).cpu().numpy().flatten()
        synthetic_100k = model.sample(100000).cpu().numpy().flatten()

    print(f"\n  Generated scenarios:")
    print(f"    1,000 samples   - Min: {synthetic_1k.min():.4f}, Max: {synthetic_1k.max():.4f}")
    print(f"    10,000 samples  - Min: {synthetic_10k.min():.4f}, Max: {synthetic_10k.max():.4f}")
    print(f"    100,000 samples - Min: {synthetic_100k.min():.4f}, Max: {synthetic_100k.max():.4f}")

    # Compare statistics
    print("\n  Statistical comparison:")
    print("  Metric      | Historical | Synthetic (100k)")
    print("  " + "-" * 45)
    print(f"  Mean        | {returns_flat.mean():+.4f}     | {synthetic_100k.mean():+.4f}")
    print(f"  Std         | {returns_flat.std():.4f}     | {synthetic_100k.std():.4f}")
    print(f"  Skewness    | {stats.skew(returns_flat):+.4f}     | {stats.skew(synthetic_100k):+.4f}")
    print(f"  Kurtosis    | {stats.kurtosis(returns_flat):+.4f}     | {stats.kurtosis(synthetic_100k):+.4f}")
    print(f"  1% quantile | {np.percentile(returns_flat, 1):+.4f}     | {np.percentile(synthetic_100k, 1):+.4f}")
    print(f"  99% quantile| {np.percentile(returns_flat, 99):+.4f}     | {np.percentile(synthetic_100k, 99):+.4f}")

    # -----------------------------------------------------------------
    # Step 4: Stress testing with synthetic data
    # -----------------------------------------------------------------
    print("\n[Step 4] Stress testing...")

    stress_scenarios = ["2x_vol", "fat_tails", "crash"]

    for scenario in stress_scenarios:
        results = stress_test(model, scenario_type=scenario, n_samples=100000)
        print(f"\n  Scenario: {scenario}")
        print(f"    Base VaR 95%:     {results['VaR_95_base']:+.4f}")
        print(f"    Stressed VaR 95%: {results['VaR_95_stressed']:+.4f}")
        print(f"    Base max loss:    {results['max_loss_base']:+.4f}")
        print(f"    Stressed max loss:{results['max_loss_stressed']:+.4f}")

    # -----------------------------------------------------------------
    # Step 5: Monte Carlo simulation for trading strategy
    # -----------------------------------------------------------------
    print("\n[Step 5] Monte Carlo simulation for trading strategy...")

    # Simple strategy: always long with 100% position
    initial_capital = 10000
    position_size = 1.0  # 100% invested
    n_simulations = 1000
    n_days = 30  # 30-day simulation

    # Generate return paths
    print(f"\n  Simulating {n_simulations} paths of {n_days} days each...")

    with torch.no_grad():
        all_returns = model.sample(n_simulations * n_days).cpu().numpy().flatten()

    return_matrix = all_returns.reshape(n_simulations, n_days)

    # Calculate portfolio values
    cumulative_returns = np.exp(np.cumsum(return_matrix, axis=1))
    final_values = initial_capital * cumulative_returns[:, -1]

    # Statistics
    mean_final = final_values.mean()
    median_final = np.median(final_values)
    worst_case = final_values.min()
    best_case = final_values.max()
    var_95 = np.percentile(final_values, 5)
    var_99 = np.percentile(final_values, 1)

    print(f"\n  Monte Carlo Results (30-day horizon):")
    print(f"    Initial capital:    ${initial_capital:,.0f}")
    print(f"    Mean final value:   ${mean_final:,.0f}")
    print(f"    Median final value: ${median_final:,.0f}")
    print(f"    Best case:          ${best_case:,.0f}")
    print(f"    Worst case:         ${worst_case:,.0f}")
    print(f"    5th percentile:     ${var_95:,.0f}")
    print(f"    1st percentile:     ${var_99:,.0f}")

    # Probability calculations
    prob_loss = (final_values < initial_capital).mean()
    prob_loss_10pct = (final_values < initial_capital * 0.9).mean()
    prob_gain_20pct = (final_values > initial_capital * 1.2).mean()

    print(f"\n  Probability estimates:")
    print(f"    P(loss):          {prob_loss*100:.1f}%")
    print(f"    P(loss > 10%):    {prob_loss_10pct*100:.1f}%")
    print(f"    P(gain > 20%):    {prob_gain_20pct*100:.1f}%")

    # -----------------------------------------------------------------
    # Step 6: Generate extreme scenarios
    # -----------------------------------------------------------------
    print("\n[Step 6] Analyzing extreme scenarios from synthetic data...")

    # Find extreme events in synthetic data
    threshold_negative = np.percentile(synthetic_100k, 0.1)  # 0.1% worst
    threshold_positive = np.percentile(synthetic_100k, 99.9)  # 0.1% best

    extreme_negative = synthetic_100k[synthetic_100k < threshold_negative]
    extreme_positive = synthetic_100k[synthetic_100k > threshold_positive]

    print(f"\n  Extreme negative events (worst 0.1%):")
    print(f"    Count: {len(extreme_negative)}")
    print(f"    Average: {extreme_negative.mean():.4f}")
    print(f"    Worst: {extreme_negative.min():.4f}")

    print(f"\n  Extreme positive events (best 0.1%):")
    print(f"    Count: {len(extreme_positive)}")
    print(f"    Average: {extreme_positive.mean():.4f}")
    print(f"    Best: {extreme_positive.max():.4f}")

    # -----------------------------------------------------------------
    # Step 7: Visualization
    # -----------------------------------------------------------------
    print("\n[Step 7] Creating visualization...")

    fig, axes = plt.subplots(2, 2, figsize=(12, 10))

    # Plot 1: Histogram comparison
    ax1 = axes[0, 0]
    ax1.hist(returns_flat, bins=50, density=True, alpha=0.6, label='Historical')
    ax1.hist(synthetic_100k, bins=100, density=True, alpha=0.4, label='Synthetic (100k)')
    ax1.set_xlabel('Return')
    ax1.set_ylabel('Density')
    ax1.set_title('Distribution Comparison')
    ax1.legend()

    # Plot 2: Monte Carlo paths
    ax2 = axes[0, 1]
    for i in range(min(100, n_simulations)):
        ax2.plot(cumulative_returns[i], alpha=0.1, color='blue')
    ax2.axhline(y=1.0, color='red', linestyle='--', label='Initial')
    ax2.set_xlabel('Day')
    ax2.set_ylabel('Cumulative Return')
    ax2.set_title(f'Monte Carlo Simulation ({n_simulations} paths)')
    ax2.legend()

    # Plot 3: Final value distribution
    ax3 = axes[1, 0]
    ax3.hist(final_values, bins=50, density=True, alpha=0.7)
    ax3.axvline(x=initial_capital, color='red', linestyle='--', label='Initial capital')
    ax3.axvline(x=var_95, color='orange', linestyle='--', label='5th percentile')
    ax3.set_xlabel('Final Portfolio Value ($)')
    ax3.set_ylabel('Density')
    ax3.set_title('Final Value Distribution')
    ax3.legend()

    # Plot 4: QQ plot
    ax4 = axes[1, 1]
    percentiles = np.linspace(0.1, 99.9, 100)
    hist_quantiles = np.percentile(returns_flat, percentiles)
    synth_quantiles = np.percentile(synthetic_100k, percentiles)
    ax4.scatter(hist_quantiles, synth_quantiles, alpha=0.6)
    ax4.plot([hist_quantiles.min(), hist_quantiles.max()],
             [hist_quantiles.min(), hist_quantiles.max()], 'r--')
    ax4.set_xlabel('Historical Quantiles')
    ax4.set_ylabel('Synthetic Quantiles')
    ax4.set_title('QQ Plot: Synthetic vs Historical')

    plt.tight_layout()

    output_path = os.path.join(os.path.dirname(__file__), 'synthetic_generation_results.png')
    plt.savefig(output_path, dpi=150)
    print(f"\n  Saved visualization to: {output_path}")

    plt.show()

    # -----------------------------------------------------------------
    # Summary
    # -----------------------------------------------------------------
    print("\n" + "=" * 60)
    print("Summary")
    print("=" * 60)
    print("""
Key capabilities demonstrated:

1. Synthetic Data Generation:
   - Generate unlimited realistic return scenarios
   - Match historical distribution characteristics
   - Preserve fat tails and skewness

2. Stress Testing:
   - Test portfolio under extreme conditions
   - Double volatility scenarios
   - Market crash simulations

3. Monte Carlo Simulation:
   - Simulate portfolio paths over time
   - Compute probability of various outcomes
   - Estimate VaR and CVaR for longer horizons

4. Practical Applications:
   - Backtest strategies on more data than available
   - Regulatory stress testing (Basel III)
   - Risk budgeting and allocation
   - Options pricing with realistic distributions
""")


if __name__ == "__main__":
    main()

File: rust_normalizing_flows/examples/synthetic_generation.rs
//! Example: Generate synthetic financial data with normalizing flows
//!
//! This example demonstrates how to:
//! 1. Train a flow on historical returns
//! 2. Generate unlimited realistic synthetic scenarios
//! 3. Use synthetic data for Monte Carlo simulation
//! 4. Perform stress testing

use normalizing_flows_finance::prelude::*;
use normalizing_flows_finance::risk::metrics::{compute_risk_metrics, stress_test};
use normalizing_flows_finance::utils::returns::{generate_synthetic_returns, standardize};
use normalizing_flows_finance::utils::statistics::describe;

fn main() -> anyhow::Result<()> {
    env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("info")).init();

    println!("═══════════════════════════════════════════════════════════");
    println!("  Synthetic Data Generation with Normalizing Flows");
    println!("═══════════════════════════════════════════════════════════");

    // Step 1: Prepare training data
    println!("\n[1] Preparing training data...");
    let original_returns = generate_synthetic_returns(500, 4.0); // Fat-tailed returns
    let orig_stats = describe(&original_returns);

    println!("    Original data:");
    println!("      Samples:  {}", orig_stats.count);
    println!("      Mean:     {:+.4}", orig_stats.mean);
    println!("      Std:      {:.4}", orig_stats.std);
    println!("      Kurtosis: {:.2}", orig_stats.kurtosis);

    let standardized = standardize(&original_returns);

    // Step 2: Train flow
    println!("\n[2] Training normalizing flow...");
    let config = FlowConfig::new(1).with_n_layers(6).with_hidden_dim(64);
    let mut flow = RealNVP::new(config);

    let _history = flow.train(&standardized, 80)?;
    println!("    Training complete!");

    // Step 3: Generate synthetic scenarios
    println!("\n[3] Generating synthetic scenarios...");

    for &n in &[1000, 10000, 100000] {
        let samples = flow.sample_flat(n);
        let stats = describe(&samples);
        println!(
            "    {:>7} samples: mean={:+.4}, std={:.4}, kurt={:.2}",
            n, stats.mean, stats.std, stats.kurtosis
        );
    }

    // Generate scaled back to original space
    let synthetic_100k_std = flow.sample_flat(100000);
    let synthetic_100k: Vec<f64> = synthetic_100k_std
        .iter()
        .map(|&x| x * orig_stats.std + orig_stats.mean)
        .collect();

    let synth_stats = describe(&synthetic_100k);
    println!("\n    Scaled to original space (100k samples):");
    println!("      Mean:     {:+.4}", synth_stats.mean);
    println!("      Std:      {:.4}", synth_stats.std);
    println!("      Kurtosis: {:.2}", synth_stats.kurtosis);

    // Step 4: Compare statistics
    println!("\n[4] Comparing distributions...");
    println!("\n    {:^15} {:>15} {:>15}", "Metric", "Original", "Synthetic");
    println!("    {}", "─".repeat(50));
    println!(
        "    {:^15} {:>15.4} {:>15.4}",
        "Mean", orig_stats.mean, synth_stats.mean
    );
    println!(
        "    {:^15} {:>15.4} {:>15.4}",
        "Std", orig_stats.std, synth_stats.std
    );
    println!(
        "    {:^15} {:>15.4} {:>15.4}",
        "Skewness", orig_stats.skewness, synth_stats.skewness
    );
    println!(
        "    {:^15} {:>15.4} {:>15.4}",
        "Kurtosis", orig_stats.kurtosis, synth_stats.kurtosis
    );
    println!(
        "    {:^15} {:>15.4} {:>15.4}",
        "Min", orig_stats.min, synth_stats.min
    );
    println!(
        "    {:^15} {:>15.4} {:>15.4}",
        "Max", orig_stats.max, synth_stats.max
    );

    // Step 5: Monte Carlo simulation
    println!("\n[5] Monte Carlo simulation for portfolio risk...");

    let initial_capital = 10000.0;
    let n_simulations = 1000;
    let n_days = 30;

    // Generate return paths
    let mut final_values = Vec::with_capacity(n_simulations);
    let mut rng_seed = 42u64;

    for _ in 0..n_simulations {
        // Generate 30 days of returns
        let path_returns: Vec<f64> = (0..n_days)
            .map(|_| {
                rng_seed = rng_seed.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
                let idx = (rng_seed as usize) % synthetic_100k.len();
                synthetic_100k[idx]
            })
            .collect();

        // Calculate cumulative return
        let cumulative: f64 = path_returns.iter().map(|r| (1.0 + r)).product();
        final_values.push(initial_capital * cumulative);
    }

    let final_stats = describe(&final_values);

    println!("\n    Monte Carlo Results ({} paths, {} days):", n_simulations, n_days);
    println!("      Initial capital:  ${:.0}", initial_capital);
    println!("      Mean final value: ${:.0}", final_stats.mean);
    println!("      Median:           ${:.0}", final_stats.median);
    println!("      5th percentile:   ${:.0}", final_stats.q1);
    println!("      95th percentile:  ${:.0}", final_stats.q3);
    println!("      Worst case:       ${:.0}", final_stats.min);
    println!("      Best case:        ${:.0}", final_stats.max);

    // Probability calculations
    let prob_loss = final_values.iter().filter(|&&v| v < initial_capital).count() as f64
        / final_values.len() as f64;
    let prob_loss_10 = final_values
        .iter()
        .filter(|&&v| v < initial_capital * 0.9)
        .count() as f64
        / final_values.len() as f64;
    let prob_gain_20 = final_values
        .iter()
        .filter(|&&v| v > initial_capital * 1.2)
        .count() as f64
        / final_values.len() as f64;

    println!("\n    Probability estimates:");
    println!("      P(any loss):     {:.1}%", prob_loss * 100.0);
    println!("      P(loss > 10%):   {:.1}%", prob_loss_10 * 100.0);
    println!("      P(gain > 20%):   {:.1}%", prob_gain_20 * 100.0);

    // Step 6: Extreme scenario analysis
    println!("\n[6] Extreme scenario analysis...");

    let threshold_1pct = synth_stats.q1; // Approximate 1st percentile
    let extreme_losses: Vec<f64> = synthetic_100k
        .iter()
        .filter(|&&x| x < threshold_1pct)
        .copied()
        .collect();

    println!(
        "    Extreme negative events (bottom 1%): {} occurrences",
        extreme_losses.len()
    );
    if !extreme_losses.is_empty() {
        let extreme_stats = describe(&extreme_losses);
        println!("      Average extreme loss: {:.2}%", extreme_stats.mean * 100.0);
        println!("      Worst loss:           {:.2}%", extreme_stats.min * 100.0);
    }

    // Step 7: Stress testing
    println!("\n[7] Stress testing scenarios...");

    let scenarios = ["2x_vol", "crash"];
    for scenario in scenarios {
        let result = stress_test(&flow, scenario, 50000);
        println!("\n    Scenario: {}", scenario);
        println!(
            "      95% VaR: {:.4} -> {:.4}",
            result.base_var_95, result.stressed_var_95
        );
        println!(
            "      Max loss: {:.4} -> {:.4}",
            result.max_loss_base, result.max_loss_stressed
        );
    }

    // Step 8: Risk metrics summary
    println!("\n[8] Flow-based risk metrics...");
    let metrics = compute_risk_metrics(&flow, 100000);

    println!("    VaR 95%:   {:.4}", metrics.var_95);
    println!("    VaR 99%:   {:.4}", metrics.var_99);
    println!("    CVaR 95%:  {:.4}", metrics.cvar_95);
    println!("    CVaR 99%:  {:.4}", metrics.cvar_99);

    // Summary
    println!("\n═══════════════════════════════════════════════════════════");
    println!("  Summary");
    println!("═══════════════════════════════════════════════════════════");
    println!();
    println!("  Capabilities demonstrated:");
    println!("  1. Generate unlimited realistic return scenarios");
    println!("  2. Match historical distribution characteristics");
    println!("  3. Monte Carlo simulation for portfolio risk");
    println!("  4. Stress testing under extreme conditions");
    println!();
    println!("  Applications:");
    println!("  - Backtesting with more data than available");
    println!("  - Regulatory stress testing (Basel III)");
    println!("  - Risk budgeting and allocation");
    println!("  - Options pricing with realistic distributions");
    println!();

    Ok(())
}
